### PR TITLE
Add table and column loading options

### PR DESCRIPTION
**Pull request by michael-yx-wu**
#### Checklist
- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:
Make it easier to set the loading state on common sections of the table such as the headers and body by specifying loading options on `Table` and `Column`, components higher up in the hierarchy. Only exposing a `loading` prop on cell components results in needing to define a `cellRenderer` function just to conditionally set the loading state. 

When `loadingOptions` are specified on both `Table` and `Column` and `loading` is also specified on `Cell`, the lowest component takes precedence. This allows users to manually make changes to the overall loading appearance of the table. 

In the following example,

```
private render() {
  <Table loadingOptions={[ "cells", "column-header" ]}>
    <Column loadingOptions={[ "cells" ]} />
    <Column />
    <Column />
  </Table>
}

private cellRenderer = (row: number, col: number) => {
  return <Cell loading={row === 0 && col === 0} />;
}
```

All column / row headers and body cells show the loadings state except:
- col1 header cell
- all col1 cells except the first one

### CODE REVIEW

**Comment by michael-yx-wu** [comment]
This is ready for review now as well 😄

File: packages/table/src/column.tsx [hunk 1]
```diff
@@ -21,6 +24,16 @@ export interface IColumnProps extends IColumnNameProps, IProps {
     id?: string | number;
 
     /**
+     * A list of `ColumnLoadingOption`. Set this prop to specify whether to render the loading state
```
Comment (giladgray) [inline packages/table/src/column.tsx]: first sentence is precisely the type--not useful to repeat.

File: packages/table/src/column.tsx [hunk 2]
```diff
@@ -21,6 +24,16 @@ export interface IColumnProps extends IColumnNameProps, IProps {
...
+     * of the column header and cells in this column. When using this in conjunction with the
+     * `loadingOptions` prop on `Table`, the options here effectively replace the options from the
+     * parent `Table` when it comes time to render cells. For example, if you set `loadingOptions=[
+     * TableLoadingOption.CELLS ]` on `Table` and `loadingOptions=[ ColumnLoadingOption.HEADER ]` on
+     * a `Column`, the cells in that column with _not_ show their loading state.
```
Comment (giladgray) [inline packages/table/src/column.tsx]: please revise this copy so it's more terse. looks like a wall of text that is hard to follow.

File: packages/table/src/headers/columnHeader.tsx [hunk 3]
```diff
@@ -40,6 +41,13 @@ export interface IColumnHeaderProps extends ISelectableProps, IColumnIndices, IC
     grid: Grid;
 
     /**
+     * If true, all `ColumnHeaderCell` render their loading state except for
```
Comment (giladgray) [inline packages/table/src/headers/columnHeader.tsx]: how many `ColumnHeaderCell`s can there be in a single `ColumnHeader`?

File: packages/table/src/table.tsx [hunk 4]
```diff
@@ -432,13 +449,23 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
 
     private columnHeaderCellRenderer = (columnIndex: number) => {
         const props = this.getColumnProps(columnIndex);
+        let loading: boolean;
+        switch (this.hasLoadingOption(props.loadingOptions, ColumnLoadingOption.HEADER)) {
```
Comment (giladgray) [inline packages/table/src/table.tsx]: `if ... else`? switch seems quite overkill and that empty default case is painful.

Comment (giladgray) [inline packages/table/src/table.tsx]: actually wait what's wrong with assigning `loading = this.hasLoadingOption(...)` directly?

File: packages/table/src/table.tsx [hunk 5]
```diff
@@ -511,6 +541,7 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
                     grid={grid}
                     locator={locator}
                     isResizable={isRowResizable}
+                    loading={loadingOptions.indexOf(TableLoadingOption.ROW_HEADERS) !== -1}
```
Comment (giladgray) [inline packages/table/src/table.tsx]: I find `>= 0` to be a more readable check for `indexOf`

File: packages/table/src/table.tsx [hunk 6]
```diff
@@ -530,14 +561,34 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
     }
 
     private bodyCellRenderer = (rowIndex: number, columnIndex: number) => {
-        return this.getColumnProps(columnIndex).renderCell(rowIndex, columnIndex);
+        const columnProps = this.getColumnProps(columnIndex);
+        const cell = columnProps.renderCell(rowIndex, columnIndex);
+        const { loading: cellLoading } = cell.props;
```
Comment (giladgray) [inline packages/table/src/table.tsx]: _whoa_ i did not know this was valid syntax! since you're only deconstructing one prop, i'd just do the simple legible `const cellLoading = cell.props.loading`.

File: packages/table/src/table.tsx [hunk 7]
```diff
@@ -530,14 +561,34 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+            loading = cellLoading;
+        } else {
+            switch (this.hasLoadingOption(columnProps.loadingOptions, ColumnLoadingOption.CELLS)) {
+                case true:
+                    loading = true;
+                    break;
+                case false:
+                    loading = false;
+                    break;
+                default:
```
Comment (giladgray) [inline packages/table/src/table.tsx]: again with the scary switch...

File: packages/table/src/regions.ts [hunk 8]
```diff
@@ -32,6 +37,11 @@ export enum RegionCardinality {
      * A region that represents all cells in the table.
      */
     FULL_TABLE,
+
+    /**
+     * A region that represents all row header cells.
+     */
+    ROW_HEADERS,
```
Comment (giladgray) [inline packages/table/src/regions.ts]: :stop_sign: are these new values supported everywhere `RegionCardinality` is used?

File: packages/table/src/table.tsx (hunk 6 above)
Comment (michael-yx-wu) [inline packages/table/src/table.tsx]: will do. and yup, it's pretty convenient when implementing `shouldComponentUpdate`

File: packages/table/src/headers/columnHeader.tsx (hunk 3 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/columnHeader.tsx]: oops, just one

File: packages/table/src/regions.ts (hunk 8 above)
Comment (michael-yx-wu) [inline packages/table/src/regions.ts]: oops, removing. see below

File: packages/table/src/headers/columnHeader.tsx (hunk 3 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/columnHeader.tsx]: double oops. `ColumnHeader` is the container for all `ColumnHeaderCell` so to answer your question: as many columns as there are in the table

File: packages/table/src/table.tsx (hunk 4 above)
Comment (michael-yx-wu) [inline packages/table/src/table.tsx]: `this.hasLoadingOptions` returns `undefined` is the first argument is `null` or `undefined`. that the `loading` variable can be `undefined` is important. where `false` means "don't show the loading state", `undefined` means defer to the table loading options

Comment (michael-yx-wu) [inline packages/table/src/table.tsx]: am i abusing `undefined` here?

File: packages/table/examples/columnLoadingExample.tsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        for (let i = 0; i < numColumns; i++) {
+            options.push(
+                <option key={i} value={i}>
+                    {this.formatColumnName(Object.getOwnPropertyNames(firstSpaceRock)[i])}
+                </option>,
+            );
+        }
+        return (
+            <label className={Classes.LABEL}>
+                Loading Column
```
Comment (llorca) [inline packages/table/examples/columnLoadingExample.tsx]: Sentence case: `Loading column`

Comment (michael-yx-wu) [inline packages/table/examples/columnLoadingExample.tsx]: oops!

File: packages/table/src/column.tsx [hunk 10]
```diff
@@ -21,6 +24,16 @@ export interface IColumnProps extends IColumnNameProps, IProps {
     id?: string | number;
 
     /**
+     * Set this prop to specify whether to render the loading state of the
+     * column header and cells * in this column. Column-level `loadingOptions`
+     * override Table-level `loadingOptions`. For example, if you set
+     * `loadingOptions=[ TableLoadingOption.CELLS ]` on `Table` and
+     * `loadingOptions=[ ColumnLoadingOption.HEADER ]` on a `Column`, the cells
+     * in that column with _not_ show their loading state.
```
Comment (giladgray) [inline packages/table/src/column.tsx]: typo: `will _not_`

File: packages/table/src/column.tsx [hunk 11]
```diff
@@ -31,10 +44,11 @@ export interface IColumnProps extends IColumnNameProps, IProps {
      * index and returns a `ColumnHeaderCell` React element
      */
     renderColumnHeader?: IColumnHeaderRenderer;
+
```
Comment (giladgray) [inline packages/table/src/column.tsx]: ❌  whitespace

File: packages/table/src/table.tsx (hunk 4 above)
Comment (giladgray) [inline packages/table/src/table.tsx]: how is that not the same as this? if `this.hasLoadingOption()` returns `undefined` then `let loading` never gets set so it is also `undefined`. in my proposal you at least intentionally assign `undefined` to it.

File: packages/table/src/table.tsx [hunk 12]
```diff
@@ -548,10 +599,12 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         const columnIndices = grid.getColumnIndicesInRect(viewportRect, fillBodyWithGhostCells);
         const noVerticalScroll = fillBodyWithGhostCells &&
             grid.isGhostIndex(rowIndices.rowIndexEnd, 0) &&
-            viewportRect != null && viewportRect.top === 0;
+            viewportRect != null && viewportRect.top === 0 ||
+            this.hasLoadingOption(loadingOptions, TableLoadingOption.ROW_HEADERS);
         const noHorizontalScroll = fillBodyWithGhostCells &&
             grid.isGhostIndex(0, columnIndices.columnIndexEnd) &&
-            viewportRect != null && viewportRect.left === 0;
+            viewportRect != null && viewportRect.left === 0 ||
```
Comment (giladgray) [inline packages/table/src/table.tsx]: boolean logic question: is `A & B & C || D` equivalent to `(A & B & C) || D`? 

just making sure that this equation is correct, as you intended it.

File: packages/table/src/tableBody.tsx [hunk 13]
```diff
@@ -32,6 +32,12 @@ export interface ITableBodyProps extends ISelectableProps, IRowIndices, IColumnI
     grid: Grid;
 
     /**
+     * If true, all `Cell` render their loading state except for those who have
```
Comment (giladgray) [inline packages/table/src/tableBody.tsx]: happens a few times in this PR: "all singular" sounds wrong. i usually write "all `Cell`s" to follow english grammar rules.

File: packages/table/test/loadingOptionsTests.tsx [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        ];
+        const tableHarness = harness.mount(
+            <Table loadingOptions={loadingOptions} numRows={2}>
+                <Column name="Column1" loadingOptions={[ ColumnLoadingOption.HEADER ]} renderCell={renderCell} />
+                <Column name="Column2" renderCell={renderCell} />
+            </Table>,
+        );
+        const tableElement = tableHarness.element.children[0];
+
+        expect(tableElement.textContent).to.equal("row0col1|row1col0|");
```
Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: i don't get it... how does this test the spec? 
can you leave a code comment explaining how this is intended to function?

File: packages/table/test/tableTests.tsx [hunk 15]
```diff
@@ -56,6 +56,24 @@ describe("<Table>", () => {
...
+        ];
+        const tableHarness = harness.mount(
+            <Table loadingOptions={loadingOptions} numRows={2}>
+                <Column name="Column0" renderCell={renderCell} />
+                <Column name="Column1" renderCell={renderCell} />
+            </Table>,
+        );
+        const tableElement = tableHarness.element.children[0];
+
+        expect(tableElement.textContent).to.equal("");
```
Comment (giladgray) [inline packages/table/test/tableTests.tsx]: and again, why is empty table content expected here?

**Comment by michael-yx-wu** [comment]
i'll be changing the target branch of this PR to `master` once #452 merges

File: packages/table/src/table.tsx (hunk 12 above)
Comment (michael-yx-wu) [inline packages/table/src/table.tsx]: yup, relying on `&` taking priority over `||`

File: packages/table/src/tableBody.tsx (hunk 13 above)
Comment (michael-yx-wu) [inline packages/table/src/tableBody.tsx]: i'll hunt around for some other instances of this

File: packages/table/test/tableTests.tsx (hunk 15 above)
Comment (michael-yx-wu) [inline packages/table/test/tableTests.tsx]: the cells and column headers and row headers all should have text content. this is testing that all three of the table loading options are working correctly and causing all headers and cells to render loading state instead of content.

File: packages/table/test/loadingOptionsTests.tsx (hunk 14 above)
Comment (michael-yx-wu) [inline packages/table/test/loadingOptionsTests.tsx]: will do!

Comment (michael-yx-wu) [inline packages/table/test/loadingOptionsTests.tsx]: this behavior is also described in the `Loading` section in the docs. let me know if that section reads unclear as well

File: packages/table/src/table.tsx (hunk 4 above)
Comment (michael-yx-wu) [inline packages/table/src/table.tsx]: i'll have to have something like this:

```
let hasLoading = this.hasLoadingOption(...);
if (hasLoading) {
  loading = true;
} else if (hasLoading === false) {
  loading = false;
}
```

Comment (michael-yx-wu) [inline packages/table/src/table.tsx]: hm it looks a little better than the switch

File: packages/table/src/_docs.scss [hunk 16]
```diff
@@ -128,17 +128,54 @@ Loading states
...
+
+Weight: 4
+
+Styleguide components.table-js.loading
+*/
+
+/*
+Table
+
+`Table` expose a `loadingOptions` prop that allows you to control the loading state behavior of all
```
Comment (giladgray) [inline packages/table/src/_docs.scss]: "exposes"

please do some grammar review of your sentences so i don't have to.

File: packages/table/src/_docs.scss [hunk 17]
```diff
@@ -128,17 +128,54 @@ Loading states
...
+
+@react-example TableLoadingExample
+
+Styleguide components.table-js.loading.table
+*/
+
+/*
+Columns
+
+`Column` also expose a `loadingOptions` prop that allows you to control the loading state behavior
```
Comment (giladgray) [inline packages/table/src/_docs.scss]: "exposes"

File: packages/table/src/_docs.scss [hunk 18]
```diff
@@ -128,17 +128,54 @@ Loading states
...
+of an individual columns' header cell and body cells. Try selecting a different column.
```
Comment (giladgray) [inline packages/table/src/_docs.scss]: "column's"

Comment (giladgray) [inline packages/table/src/_docs.scss]: "selecting a different column" -- is this what you meant to say?

File: packages/table/src/_docs.scss [hunk 19]
```diff
@@ -128,17 +128,54 @@ Loading states
...
+
+@react-example ColumnLoadingExample
+
+Styleguide components.table-js.loading.column
+
+Weight: 1
```
Comment (giladgray) [inline packages/table/src/_docs.scss]: Weight immediately before Styleguide (with newlines) always

File: packages/table/src/_docs.scss [hunk 20]
```diff
@@ -128,17 +128,54 @@ Loading states
...
+control of which cells should show a loading state. Try selecting a different preset loading
+configuration.
 
 @react-example CellLoadingExample
 
-Weight: 4
+Styleguide components.table-js.loading.cell
 
-Styleguide components.table-js.loading
+Weight: 2
```
Comment (giladgray) [inline packages/table/src/_docs.scss]: Weight before Styleguide

File: packages/table/src/cell/cell.tsx [hunk 21]
```diff
@@ -47,7 +47,7 @@ export interface ICellProps extends IIntentProps, IProps {
 
 export type ICellRenderer = (rowIndex: number, columnIndex: number) => React.ReactElement<ICellProps>;
 
-export const emptyCellRenderer = (_rowIndex: number, _columnIndex: number) => <Cell />;
+export const emptyCellRenderer: ICellRenderer = (_rowIndex: number, _columnIndex: number) => <Cell />;
```
Comment (giladgray) [inline packages/table/src/cell/cell.tsx]: this change shouldn't be necessary. you can even remove the args since they're not used. a function with fewer args will pass type checking.

File: packages/table/src/table.tsx [hunk 22]
```diff
@@ -432,13 +449,20 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
 
     private columnHeaderCellRenderer = (columnIndex: number) => {
         const props = this.getColumnProps(columnIndex);
+        let loading: boolean;
+        const hasLoadingOption = this.hasLoadingOption(props.loadingOptions, ColumnLoadingOption.HEADER);
+        if (hasLoadingOption) {
+            loading = true;
+        } else if (hasLoadingOption === false) {
+            loading = false;
+        }
```
Comment (giladgray) [inline packages/table/src/table.tsx]: still convinced you can do `const loading = this.hasLoadingOption(props.loadingOptions, ColumnLoadingOption.HEADER);` and it will be either `true`, `false`, or `undefined`, exactly as it is now.

have you tried it?

File: packages/table/src/table.tsx [hunk 23]
```diff
@@ -530,14 +558,31 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+        if (cellLoading != null) {
+            loading = cellLoading;
+        } else {
+            const hasLoadingOption = this.hasLoadingOption(columnProps.loadingOptions, ColumnLoadingOption.CELLS);
+            if (hasLoadingOption) {
+                loading = true;
+            } else if (hasLoadingOption === false) {
+                loading = false;
+            }
+        }
```
Comment (giladgray) [inline packages/table/src/table.tsx]: and again with the direct assignment https://github.com/palantir/blueprint/pull/473/files#r98076641

File: packages/table/test/columnTests.tsx [hunk 24]
```diff
@@ -51,4 +51,41 @@ describe("Column", () => {
...
+        expect(table.find(columnHeaderSelector, 2).text()).to.equal("Two");
+
+        const col0cells = table.element.querySelectorAll(".bp-table-cell-col-0");
+        for (let i = 0; i < col0cells.length; i++) {
+            expect(col0cells.item(i).textContent).to.equal("");
+        }
+
+        const col1cells = table.element.querySelectorAll(".bp-table-cell-col-1");
+        for (let i = 0; i < col1cells.length; i++) {
+            expect(col1cells.item(i).textContent).to.equal("");
```
Comment (giladgray) [inline packages/table/test/columnTests.tsx]: does it make more sense to test _instead_ that each of these cells contains a `.pt-skeleton`? seems more reliable and logical than asserting that empty text means they are in the loading state.

File: packages/table/test/loadingOptionsTests.tsx [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+        ];
+        const tableHarness = harness.mount(
+            <Table loadingOptions={loadingOptions} numRows={2}>
+                <Column name="Column0" loadingOptions={[ ColumnLoadingOption.HEADER ]} renderCell={renderCell} />
+                <Column name="Column1" renderCell={renderCell} />
+            </Table>,
+        );
+        const tableElement = tableHarness.element.children[0];
+
+        /**
```
Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: for the record, don't need the second `*` here as this is merely a multiline comment, not a JSDoc comment that you'd expect to appear in consumer documentation.

File: packages/table/test/tableTests.tsx (hunk 15 above)
Comment (giladgray) [inline packages/table/test/tableTests.tsx]: 👎  but it doesn't test that they __render loading content__. it tests that they __don't render their text content__, which is not _necessarily_ the same thing. do you see the difference?

File: packages/table/src/_docs.scss (hunk 16 above)
Comment (michael-yx-wu) [inline packages/table/src/_docs.scss]: I'm using `Table` in the plural sense, will change to singular

File: packages/table/test/columnTests.tsx (hunk 24 above)
Comment (michael-yx-wu) [inline packages/table/test/columnTests.tsx]: yup that's fair. i'll change the table / loading tests as well then

File: packages/table/src/_docs.scss (hunk 18 above)
Comment (michael-yx-wu) [inline packages/table/src/_docs.scss]: what do you mean?

Comment (giladgray) [inline packages/table/src/_docs.scss]: oh i get it, there's a `<select>` to choose columns. but that sentence sure sounds like you want me to click on a column to select it (ie, make it blue). clarify the wording, that you're simply asking folks to play with the option.

File: packages/table/src/_docs.scss (hunk 16 above)
Comment (giladgray) [inline packages/table/src/_docs.scss]: no you're not, it's very much singular here. if it's plural then add an `s` somewhere.

File: packages/table/test/loadingOptionsTests.tsx [hunk 26]
```diff
@@ -21,46 +43,130 @@ describe("Loading Options", () => {
...
+                subset.push(list[j]);
+            }
+        }
+        listOfSubsets.push(subset);
+    }
+
+    return listOfSubsets;
+}
+
+function testLoadingOptionOverrides(
```
Comment (michael-yx-wu) [inline packages/table/test/loadingOptionsTests.tsx]: this test describes the expected loading option override behavior for all types of cells. is it clear that this is the case, or should i leave a comment?

Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: comments are always appreciated, esp in a file with so much complex code.

i suspect there are some nice refactors that could be done to clean this up, but i recognize that you're basically out of time here :(

File: packages/table/src/table.tsx [hunk 27]
```diff
@@ -530,14 +556,26 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+        const columnProps = this.getColumnProps(columnIndex);
+        const cell = columnProps.renderCell(rowIndex, columnIndex);
+        const cellLoading = cell.props.loading;
+
+        let loading: boolean;
+        if (cellLoading != null) {
+            loading = cellLoading;
+        } else {
+            loading = this.hasLoadingOption(columnProps.loadingOptions, ColumnLoadingOption.CELLS);
+        }
```
Comment (giladgray) [inline packages/table/src/table.tsx]: i tend to prefer ternary ifs for this pattern so you can define the variable as a `const` and immediately assign the correct value.

```js
const loading = (cellLoading == null)
    ? this.hasLoadingOption(columnProps.loadingOptions, ColumnLoadingOption.CELLS)
    : cellLoading; 
```

File: packages/table/test/loadingOptionsTests.tsx [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+
+    /*
+     * Cell loading overrides column loading which in turn overrides table loading. If loading
+     * options are omitted, then the loading options of the parent component are used. Because there
+     * is no `Row` component, row header cell loading overrides table loading. Below is an
+     * exhaustive set of tests of all possible combinations of loading options.
+     */
+    allTableLoadingOptions.forEach((tableLoadingOptions: TableLoadingOption[]) => {
+        allColumnLoadingOptions.forEach((columnLoadingOptions: ColumnLoadingOption[]) => {
+            allRowLoadingOptions.forEach((rowLoadingOptions: RowLoadingOption[]) => {
```
Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: shouldn't need types on the params here. can they not be inferred from `forEach`?

File: packages/table/test/loadingOptionsTests.tsx [hunk 29]
```diff
@@ -0,0 +1,172 @@
...
+                it(`table: ${tableLoadingOptions}, column: ${columnLoadingOptions}, row: ${rowLoadingOptions}`, () => {
+                    const isCellLoading = (index: number) => {
```
Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: pull this function out (define at root level) cuz it's essentially static.

File: packages/table/test/loadingOptionsTests.tsx [hunk 30]
```diff
@@ -0,0 +1,172 @@
...
+                            <Column
+                                loadingOptions={columnLoadingOptions}
+                                renderColumnHeader={renderColumnHeader}
+                            />
+                            <Column
+                                loadingOptions={columnLoadingOptions}
+                                renderColumnHeader={renderColumnHeader}
+                            />
+                        </Table>,
+                    );
```
Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: could even define a quick helper component `<TableLoadingOptionsTester>` that does this composition, so it doesn't have to be redefined in each test run.

File: packages/table/test/loadingOptionsTests.tsx [hunk 31]
```diff
@@ -0,0 +1,172 @@
...
+
+    return listOfSubsets;
+}
+
+function testLoadingOptionOverrides(
+    cells: NodeListOf<Element>,
+    cellType: CellType,
+    cellLoading: (index: number) => boolean,
+    columnLoadingOptions: ColumnLoadingOption[],
+    tableLoadingOptions: TableLoadingOption[]) {
```
Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: put closing paren in line with open:

```js
function name(
   args,
   ...,
) {
```

File: packages/table/test/loadingOptionsTests.tsx [hunk 32]
```diff
@@ -0,0 +1,172 @@
...
+
+    // tslint:disable-next-line:prefer-for-of
```
Comment (giladgray) [inline packages/table/test/loadingOptionsTests.tsx]: shouldn't need to disable this? seems like valid usage.

File: packages/table/test/tableTests.tsx [hunk 33]
```diff
@@ -56,6 +58,38 @@ describe("<Table>", () => {
...
+        for (let i = 0; i < cells.length; i++) {
+            expectCellLoading(cells.item(i), CellType.BODY_CELL);
+        }
+
+        const columnHeaders = tableHarness.element.querySelectorAll(".bp-table-column-headers .bp-table-header");
+        for (let i = 0; i < columnHeaders.length; i++) {
+            expectCellLoading(columnHeaders.item(i), CellType.COLUMN_HEADER);
+        }
+
+        const rowHeaders = tableHarness.element.querySelectorAll(".bp-table-row-headers .bp-table-header");
```
Comment (giladgray) [inline packages/table/test/tableTests.tsx]: you could use `.queryAll()` instead of `querySelectorAll()` (from `dom4` polyfill) which returns an Array, then simply do `rowHeaders.forEach(...)`

File: packages/table/test/cellTestUtils.ts [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+        if (cellType !== CellType.BODY_CELL) {
+            const headerNameText = cellType === CellType.COLUMN_HEADER
+                ? cell.querySelector(".bp-table-column-name-text")
+                : cell.querySelector(".bp-table-row-name");
+            expect(headerNameText).to.not.be.null;
+            expect(headerNameText.textContent).to.equal("");
+            expect(headerNameText.children.length).to.equal(1);
+        } else {
+            expect(cell.children.length).to.equal(1);
+            expect(cell.querySelector(`.${Classes.SKELETON}`)).to.not.be.null;
```
Comment (giladgray) [inline packages/table/test/cellTestUtils.ts]: this is identical to line 24

File: packages/table/test/cellTestUtils.ts [hunk 35]
```diff
@@ -0,0 +1,40 @@
...
+    if (loading) {
+        expect(cell.classList.contains(Classes.LOADING)).to.be.true;
+        expect(cell.querySelector(`.${Classes.SKELETON}`)).to.not.be.null;
+        if (cellType !== CellType.BODY_CELL) {
+            const headerNameText = cellType === CellType.COLUMN_HEADER
+                ? cell.querySelector(".bp-table-column-name-text")
+                : cell.querySelector(".bp-table-row-name");
+            expect(headerNameText).to.not.be.null;
+            expect(headerNameText.textContent).to.equal("");
+            expect(headerNameText.children.length).to.equal(1);
```
Comment (giladgray) [inline packages/table/test/cellTestUtils.ts]: what is the one child? `.pt-skeleton`? can you make that explicit?

File: packages/table/test/loadingOptionsTests.tsx (hunk 26 above)
Comment (michael-yx-wu) [inline packages/table/test/loadingOptionsTests.tsx]: i refactored a bit before i put this commit up. i think this is probably easiest to read version of this function.

File: packages/table/test/cellTestUtils.ts [hunk 36]
```diff
@@ -21,17 +21,17 @@ export const CellType = {
...
                 ? cell.querySelector(".bp-table-column-name-text")
                 : cell.querySelector(".bp-table-row-name");
             expect(headerNameText).to.not.be.null;
             expect(headerNameText.textContent).to.equal("");
             expect(headerNameText.children.length).to.equal(1);
+            expect(headerNameText.children[0].classList.contains(Classes.SKELETON)).to.be.true;
         } else {
             expect(cell.children.length).to.equal(1);
-            expect(cell.querySelector(`.${Classes.SKELETON}`)).to.not.be.null;
+            expect(cell.children[0].classList.contains(Classes.SKELETON)).to.not.be.null;
```
Comment (giladgray) [inline packages/table/test/cellTestUtils.ts]: wait why back to this format?
did my comments lead us in a circle? 😶 

you're testing too much here and it's not helping. 
it should be sufficient to assert that a selector exists, for example:
`.bp-table-row-name .pt-skeleton`

File: packages/table/test/cellTestUtils.ts (hunk 35 above)
Comment (michael-yx-wu) [inline packages/table/test/cellTestUtils.ts]: added a test that the child is a skeleton

File: packages/table/test/cellTestUtils.ts (hunk 36 above)
Comment (michael-yx-wu) [inline packages/table/test/cellTestUtils.ts]: i wanted to explicit about where the skeleton exists, hence the single child test. it's nested further down in header cells vs. normal cells

Comment (michael-yx-wu) [inline packages/table/test/cellTestUtils.ts]: L31 and L34 are slightly different in this sense.

**Comment by giladgray** [comment]
😅 let's do this!
